File: OI/Italy/OIS21R1/keygen/A.py
#!/usr/bin/env python3

from sys import stdin, stdout

def input(): return stdin.readline().strip()
def read_int_list(): return list(map(int, input().split()))
def read_int_tuple(): return tuple(map(int, input().split()))
def read_int(): return int(input())

from itertools import product

### CODE HERE

def ok(assignment, S):
    S = S.replace('!', '_')
    env = dict()
    for i, v in enumerate(assignment):
        env[chr(ord('a') + i)] = v
        env['_' + chr(ord('a') + i)] = 1-v

    return 1 == (eval(S, env))

def ans(K, S):
    ret = 0
    for assignment in product([0,1], repeat=K):
        if ok(assignment, S):
            ret += 1
    return ret

for _ in range(read_int()):
    args = read_int(), input()
    print(ans(*args))

File: OI/Italy/OIS21R1/keygen/A.cpp
#include <bits/stdc++.h>
using namespace std;

int eval(string& S, int assignments) {
  int ret = 0;
  int curr = 1;
  int mask = 0;
  for (int i=0; i<S.size(); i++) {
    char s = S[i];
    if ((s == '^') || (s == '&') || (s == '(')) continue;
    if (s == ')') {
      ret ^= curr;
      curr = 1;
      continue;
    }
    if (s == '!') {
      mask = 1;
      continue;
    }
    s -= 'a';
    assert(0 <= s && s < 15);
    curr &= ((bool) (assignments & (1 << s))  ^ mask);
    mask = 0;
  }
  return ret;
}

int ans(int K, string& S) {
  int ret = 0;
  for (int i=0; i<(1<<K); i++) {
    ret += eval(S, i);
  }
  return ret;
}

int main() {

  int T = 0;
  cin >> T;
  while (T --> 0) {
    int K;
    string S;
    cin >> K >> S;
    cout << ans(K, S) << endl;
  }

  return 0;
}
